### PR TITLE
resource/aws_kinesis_firehose_delivery_stream: Fix for s3 destinations

### PR DESCRIPTION
**Pull request by ApsOps**
- import doesn't work for s3 destination

Fixes the breaking functionality caused by #2082

### CODE REVIEW

**Comment by dennisatspaceape** [comment]
Works for me on a newly created Delivery Stream, also verified that a subsequent `plan` does not flag any changes

**Comment by radeksimko** [comment]
Hi @ApsOps 
thanks for the PR. We're trying to get 1.7.1 (bugfix release) out this week so I took the liberty to make some tweaks and push them to your branch directly, instead of following the standard procedure (reviewing & discussing & letting you implement). I hope you don't mind.

I didn't pile up just random commits. The guiding line for me were failing acceptance tests which we already have in place. The goal was to get them all pass again which I managed to do with _all_ attached changes.

I hope my commit messages are self-explanatory - if not I'm happy to explain/update.

There's still [one pending bug](https://github.com/terraform-providers/terraform-provider-aws/issues/3005) in this resource, but it's apparently not covered in any test, so we can address it in a separate PR.

## Test results

```
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3InvalidParameterName
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3InvalidParameterName (1.68s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3InvalidProcessorType
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3InvalidProcessorType (1.77s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_s3basic
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3basic (111.15s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_s3KinesisStreamSource
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3KinesisStreamSource (120.53s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_importBasic
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_importBasic (267.21s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3basic
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3basic (298.13s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_s3WithCloudwatchLogging
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3WithCloudwatchLogging (300.28s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_missingProcessingConfiguration
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_missingProcessingConfiguration (306.58s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3Updates
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ExtendedS3Updates (314.55s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_s3ConfigUpdates
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_s3ConfigUpdates (396.72s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_RedshiftConfigUpdates
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_RedshiftConfigUpdates (922.33s)
=== RUN   TestAccAWSKinesisFirehoseDeliveryStream_ElasticsearchConfigUpdates
--- PASS: TestAccAWSKinesisFirehoseDeliveryStream_ElasticsearchConfigUpdates (1158.73s)
```

**Comment by ApsOps** [comment]
@radeksimko Thanks so much!

It would be awesome if we could have acceptance tests running on travis. I think I wasn't able to get them working properly, and it costs more if tests fail which leaves some resources lying around sometimes.

File: aws/resource_aws_kinesis_firehose_delivery_stream.go [hunk 1]
```diff
@@ -203,18 +203,43 @@ func flattenFirehoseS3Configuration(s3 firehose.S3DestinationDescription) []map[
...
+		t := *p.Type
+		parameters := make([]interface{}, 0)
+
 		for _, params := range p.Parameters {
-			processors[i]["parameters"] = map[string]interface{}{
-				"parameter_name":  params.ParameterName,
-				"parameter_value": params.ParameterValue,
+			name, value := *params.ParameterName, *params.ParameterValue
+
+			if t == "Lambda" {
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply aws/resource_aws_kinesis_firehose_delivery_stream.go]: Fixed.

File: aws/resource_aws_kinesis_firehose_delivery_stream.go [hunk 2]
```diff
@@ -269,31 +298,36 @@ func flattenKinesisFirehoseDeliveryStream(d *schema.ResourceData, s *firehose.De
 			elasticsearchConfList[0] = elasticsearchConfiguration
 			d.Set("elasticsearch_configuration", elasticsearchConfList)
 			d.Set("s3_configuration", flattenFirehoseS3Configuration(*destination.ElasticsearchDestinationDescription.S3DestinationDescription))
-		} else if destination.S3DestinationDescription != nil {
+		} else if d.Get("destination").(string) == "s3" {
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply aws/resource_aws_kinesis_firehose_delivery_stream.go]: As discussed on Slack we will eventually phase out this field as it likely doesn't need to be there at all.

**Comment by pecigonzalo** [comment]
Could it be related to https://github.com/terraform-providers/terraform-provider-aws/issues/2997 ?

**Comment by ApsOps** [comment]
@pecigonzalo Yes, this should fix #2997

**Comment by bflad** [comment]
This has been released in [terraform-provider-aws version 1.7.1](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#171-january-19-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

Please note we are tracking one additional crash relating to the (lack of) `prefix` of in a `s3_configuration` in #3071. Sorry for the trouble with this resource being quite complex and the acceptance testing quite lengthy.

**Pull request merged**